File: src/chi_testbench_gem5_v2/sequences/false_sharing.cc
/*
 * Copyright (c) 2026 ripopov
 * SPDX-License-Identifier: BSD-3-Clause
 */

#include "base/trace.hh"
#include "chi_testbench_gem5_v2/driver.hh"
#include "chi_testbench_gem5_v2/sequence_context.hh"
#include "chi_testbench_gem5_v2/sequences/registry.hh"
#include "debug/ChiTestbenchGem5V2.hh"

namespace gem5
{
namespace chi_gem5tb_v2
{
namespace
{

// Two or more tiles each write a distinct byte in the same cache line.
// The CHI protocol still treats this as a coherence conflict — every
// write from one tile forces a snoop of the other tile(s).
void
false_sharing_seq(SequenceContext &ctx)
{
    auto &drv = ctx.drv;
    const auto &p = drv.params();
    const uint64_t addr = p.line_addr + p.byte_offset;
    const uint32_t iters = p.iterations;

    DPRINTF(ChiTestbenchGem5V2, "%s false_sharing: addr=%#llx iters=%u\n",
            drv.name(), (unsigned long long)addr, iters);

    uint8_t byte = 0;
    for (uint32_t i = 0; i < iters; i++) {
        byte = static_cast<uint8_t>(i & 0xFF);
        drv.write_unique_full(addr, &byte, 1);
    }
}

[[maybe_unused]] Registrar _r("false_sharing", &false_sharing_seq);

} // namespace
} // namespace chi_gem5tb_v2
} // namespace gem5

File: ruby-book/final/chi_testbench_gem5/scenarios/false_sharing.py
# Copyright (c) 2026 ripopov
# SPDX-License-Identifier: BSD-3-Clause
"""
false_sharing — Tiles 0 and 15 each write a different byte of the same
cache line, repeatedly. Semantically disjoint bytes, but every write
by one tile invalidates the other's copy.
"""

from m5.objects import (
    ChiGem5Barrier,
    ChiSeqDriver,
    FalseSharingSequence,
)


def build(args, planner):
    line_addr = planner.address_for_hnf(hnf_idx=7, line_offset=0)
    barrier = ChiGem5Barrier(expected=2)
    iterations = max(args.scenario_iterations, 1000)

    drivers = [None] * args.num_cpus
    drivers[0] = ChiSeqDriver(
        tile_id=0,
        finish_barrier=barrier,
        sequence=FalseSharingSequence(
            line_addr=line_addr,
            byte_offset=0,
            iterations=iterations,
        ),
    )
    drivers[15] = ChiSeqDriver(
        tile_id=15,
        finish_barrier=barrier,
        sequence=FalseSharingSequence(
            line_addr=line_addr,
            byte_offset=32,
            iterations=iterations,
        ),
    )
    for tile in range(args.num_cpus):
        if drivers[tile] is None:
            drivers[tile] = ChiSeqDriver(tile_id=tile)
    return drivers
